sphinx/transforms/post_transforms/images.py

```diff
@@ -0,0 +1,92 @@
+# -*- coding: utf-8 -*-
+"""
+    sphinx.transforms.post_transforms.images
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+    Docutils transforms used by Sphinx.
+
+    :copyright: Copyright 2007-2017 by the Sphinx team, see AUTHORS.
+    :license: BSD, see LICENSE for details.
+"""
+
+import os
+
+from six import text_type
+from docutils import nodes
+
+from sphinx.transforms import SphinxTransform
+from sphinx.util import logging, requests
+from sphinx.util.osutil import ensuredir
+
+if False:
+    # For type annotation
+    from typing import Any, Dict  # NOQA
+    from sphinx.application import Sphinx  # NOQA
+
+
+logger = logging.getLogger(__name__)
+
+
+class BaseImageConverter(SphinxTransform):
+    def apply(self):
+        # type: () -> None
+        for node in self.document.traverse(nodes.image):
+            if self.match(node):
+                self.handle(node)
+
+    def match(self, node):
+        # type: (nodes.Node) -> bool
+        return True
+
+    def handle(self, node):
+        # type: (nodes.Node) -> None
+        pass
+
+
+class ImageDownloader(BaseImageConverter):
+    default_priority = 100
+
+    def match(self, node):
+        # type: (nodes.Node) -> bool
+        if self.app.builder.supported_remote_images:
+            return False
+        else:
+            return '://' in node['uri']
+
+    def handle(self, node):
+        # type: (nodes.Node) -> None
+        imgdir = os.path.join(self.app.doctreedir, 'images')
```

Comment:
I used `doctreedir` for the download directory for remote images.
I feel this is not good way, but I could not find a better directory.